x509/revoked.go

```diff
@@ -0,0 +1,358 @@
+				return nil, fmt.Errorf("x509: failed to unmarshal X.509 authority information: %v", err)
+			} else if len(rest) != 0 {
+				return nil, errors.New("x509: trailing data after X.509 authority information")
+			}
+
+			for _, v := range aia {
+				// GeneralName: uniformResourceIdentifier [6] IA5String
+				if v.Location.Tag != 6 {
+					continue
+				}
+				if v.Method.Equal(oidAuthorityInfoAccessOcsp) {
+					certList.TBSCertList.OCSPServer = append(certList.TBSCertList.OCSPServer, string(v.Location.Bytes))
+				} else if v.Method.Equal(oidAuthorityInfoAccessIssuers) {
+					certList.TBSCertList.IssuingCertificateURL = append(certList.TBSCertList.IssuingCertificateURL, string(v.Location.Bytes))
+				}
+				// TODO(drysdale): cope with more possibilities
+			}
+		default:
+			if e.Critical {
+				return nil, fmt.Errorf("unhandled critical extension in revokedCertificate: %v", e.Id)
+			}
+		}
+	}
+
+	return &certList, nil
+}
+
+func parseIssuingDistributionPoint(data []byte, idp *IssuingDistributionPoint, name *GeneralNames) error {
+	// RFC 5280 s5.2.5
+	if rest, err := asn1.Unmarshal(data, idp); err != nil {
+		return fmt.Errorf("x509: failed to unmarshal X.509 CRL issuing-distribution-point: %v", err)
+	} else if len(rest) != 0 {
+		return errors.New("x509: trailing data after X.509 CRL issuing-distribution-point")
+	}
+
+	typeCount := 0
+	if idp.OnlyContainsUserCerts {
+		typeCount++
+	}
+	if idp.OnlyContainsCACerts {
+		typeCount++
+	}
+	if idp.OnlyContainsAttributeCerts {
+		typeCount++
+	}
+	if typeCount > 1 {
+		return fmt.Errorf("x509: multiple cert types set in issuing-distribution-point: user:%v CA:%v attr:%v",
+			idp.OnlyContainsUserCerts, idp.OnlyContainsCACerts, idp.OnlyContainsAttributeCerts)
+	}
+	fnData := idp.DistributionPoint.FullName.FullBytes
+	if len(fnData) > 0 {
+		// Replace the leading context-specific tag [0] with the SEQUENCE tag.
+		data := make([]byte, len(fnData))
+		copy(data, fnData)
+		data[0] = 0x30
+		err := parseGeneralNames(data, name)
+		if err != nil {
+			return fmt.Errorf("x509: failed to parse X.509 CRL issuing-distribution-point fullName: %v", err)
+		}
+	}
+	return nil
+}
+
+// RevokedCertificate represents the unnamed ASN.1 structure that makes up the
+// revokedCertificates member of the TBSCertList structure from RFC 5280, s5.1.
+// It has the same content as pkix.RevokedCertificate but the extensions are
+// included in a parsed format.
+type RevokedCertificate struct {
+	pkix.RevokedCertificate
+	// Cracked out extensions:
+	RevocationReason RevocationReasonCode
+	InvalidityDate   time.Time
+	Issuer           GeneralNames
+}
+
+func parseRevokedCertificate(pkixRevoked pkix.RevokedCertificate) (*RevokedCertificate, error) {
+	result := RevokedCertificate{RevokedCertificate: pkixRevoked}
+	for _, e := range pkixRevoked.Extensions {
+		if expectCritical, present := certExtCritical[e.Id.String()]; present {
+			if e.Critical && !expectCritical {
```

Comment:
Are these `e` the same type as the `e` up above at L184? Worth factoring this check out if so?

Reply:
They are both the same type (`Extension`) but the errors become distinct in a later commit...